core/tomtom.lua

```diff
@@ -0,0 +1,34 @@
+-------------------------------------------------------------------------------
+---------------------------------- NAMESPACE ----------------------------------
+-------------------------------------------------------------------------------
+local _, ns = ...
+
+-------------------------------------------------------------------------------
+----------------------------------- TOM TOM -----------------------------------
+-------------------------------------------------------------------------------
+
+ns.tomtom = {}
+
+function ns.tomtom:AddSingleWaypoint(attrs)
```

Comment:
This style of function/method seems odd to me, although I see how it works. You have defined the methods using the `:` syntax, effectively making `attrs` the `self` instance, then you are calling these methods using the `.` syntax and passing in a new table as `self`. The code would work the same if you removed the `attrs` parameter entirely and just defined the method as `()` since `attrs` is getting mapped to `self` anyways.

This use of `self` as a container of the arguments passed into the function makes the interface to the function unclear. I would rather see `(node, mapID, coord)` as the signature of the function here. The `self` keyword should only be used in OO contexts to reference the current instance of a class.

Reply:
When I wrote the original function for drawing a circle POI you rewrote it to take a single object instead of multiple arguments which is why I wrote it as `attrs` instead of `(node, mapID, coord)`. I'm happy to split it back out!

Perhaps part of my misunderstanding is that I don't really know lua that well. I come from a JavaScript background which is much different.

Are you saying the first argument in a function is always `self`? If so, that would explain why I was having difficulty getting all the arguments sent through when I had them separated. Wouldn't the function have to take `(self, node, mapID, coord)` to compensate?

Reply:
I don't remember rewriting anything for the circle POI. I see this commit from you in the PR that added those new POI types: https://github.com/zarillion/handynotes-plugins/pull/178/commits/f01644065dad700f85fcca8b3c193852ea35038d

Did you not add them in using `attrs` from the start? The `Circle` and `Square` functions must have a call signature of `attrs` because that is the call signature of all POIs in general. Most POIs are subclasses of the `POI` class, but those two are plain functions so I may need to look at those again to understand how they work and if they need to be reworked as proper POI subclasses.

Functions in lua behave differently depending on whether or not you define them with `:` and call them with `:`. In general the `:` syntax should only be used for OO programming: https://www.lua.org/pil/16.html

Reply:
Hmm okay. Well what do I need to do to fix it? Just change back to `(node, mapID, coord)` for sending the arguments?